src/utils/debounce.ts

```diff
@@ -0,0 +1,40 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See LICENSE.md in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+type DisposableLike = {
+    dispose: () => any
+}
+
+const activeDebounces: { [id: string]: DisposableLike } = {};
+
+/*
+ *  To be used in validating user input through VS Code's showInputBox API
+ *  VS Code's validateInput callback can throw a silent timeout error when the delay is too long and defaults to a return of 'undefined' (causing false returns)
+ *  This function defaults to 1000ms for safety/consistency.
+ */
+export async function inputBoxDebounce<T>(delay: 500 | 750 | 1000, id: string, callback: (...args: unknown[]) => Promise<T> | T, ...args: unknown[]): Promise<T> {
```

Comment:
So I need some feedback here.  I ran into a weird issue where if a validateInput callback function has a delay that's too long, VS Code will throw a timeout error (thanks to @nturinski for helping me find that error).  VS Code seems to silently catch this error and then resolve validateInput to a false positive (I think `undefined`).

I do consider this to be a sort of janky behavior that we should avoid if possible, but the thing is, it doesn't really impact us all that much because false positives on intermediate keystrokes don't actually matter.  Only the last keystroke matters, and that works just fine.

I'd like to avoid the silent errors if possible, but it makes our code a little bit more complex.  I will post the alternate code below, but we can resolve to say undefined for intermediate keystrokes instead whenever a new keystroke is pressed, basically helping to alleviate delay buildup in intermediate keystrokes; however, now undefined will essentially become a `truthy` value for us which I find a little unintuitive.

Anyway, thoughts on which version to implement.

We will call this PR's code version A, this comments reference version B, and version C can be some entirely other implementation suggested by you.

Version B:
![image](https://user-images.githubusercontent.com/40250218/200921672-7ec6764a-e3bc-47e6-95f9-192607abfc71.png)

Example of calling it:

Version A:
![image](https://user-images.githubusercontent.com/40250218/200921790-fa9a4fc0-f7c8-44cd-ac6d-b5a39f4f0d62.png)


Version B:
![image](https://user-images.githubusercontent.com/40250218/200921749-621c6d15-cfc9-46a6-b6d6-d6389953da3a.png)

Reply:
@nturinski can you think of a way to use the `InputBox`'s `onDidAccept` event to deal with slow validation steps like this? We'd want a way to avoid prompting twice (or else the framework gets confused about back buttons and such)--so I don't think putting a loop around it is the way to go. However, _within_ the wizard/UI framework, we could add a method for these slow validation steps that checks on accept, and forces the user to try again if needed, but internally knows that this is really one prompt.

Reply:
I'm imagining an `async`, slow validation task that is checked with `onDidAccept`, and if it fails it reopens the prompt, noting the error, and supplying what the user entered as the already-inputted text (so they don't have to retype everything).

Reply:
I'll open an issue to investigate this further in `AzureTools`.  I would assume that either debounce implementation would still benefit from having a solution like the one Brandon mentioned above.

Reply:
Resume discussions here:  https://github.com/microsoft/vscode-azuretools/issues/1275

Reply:
> I'm imagining an `async`, slow validation task that is checked after `onDidAccept`, and if it fails it reopens the prompt, noting the error, and supplying what the user entered as the already-inputted text (so they don't have to retype everything).

I think that this would work. We could make the slow validation task property optional as well so that this wouldn't be a breaking change.